### ISSUE TITLE
`strict` warning shown unexpected

### ISSUE THREAD

**Issue by leftstick** [eslint/eslint#5239]
I am using `eslint` `2.0.0`

My lint config as following:

```javascript
{
    "parserOptions":{
        "ecmaVersion": 6,
        "sourceType": "script",
        "ecmaFeatures": {
            "globalReturn": false,
            "impliedStrict": false,
            "jsx": false,
            "experimentalObjectRestSpread": true
        }
    },
    "env": {
        "shared-node-browser": true,
        "commonjs": true,
        "node": true,
        "es6": true,
        "mocha": true,
        "jquery": true
    },
    "rules": {
        "strict": [1, "safe"]
    }
}
```

And my code is very simple: 
```javascript
'use strict';

module.exports = { apiRoot: 'http://localhost:3000/' };
```

According to the [strict-safe-doc](http://eslint.org/docs/rules/strict#safe-default), 

> When the node or commonjs environments are enabled or globalReturn is enabled in ecmaFeatures, ESLint considers code to be inside the module wrapper, and "safe" mode corresponds to "global" mode and enforces global strict mode directives. Everywhere else, "safe" mode corresponds to "function" mode and enforces strict mode directives inside top-level functions.

I think it should be treated as a `commonjs` module, and `global strict` directive is fine. 

But i got warning:

```
1:1  warning  Use the function form of 'use strict'  strict
```

Am i wrong in the configuration? Why should i `use function form of 'use strict'`?

**Comment by BYK** [comment]
This looks like a bug to me since the current code only looks at `globalReturn` and doesn't care about `commonjs`:

```js
mode = ecmaFeatures.globalReturn ? "global" : "function";
```

This might be an oversight. @nzakas thoughts?

**Comment by btmills** [comment]
The `commonjs` environment enables `globalReturn`, but the config file is explicitly overriding that setting. @leftstick does the warning still show if you remove the line that disables `globalReturn`?

**Comment by leftstick** [comment]
Thanks @BYK @btmills 
The warning disappeared while explicitly set `globalReturn` to `true`.

Then i got another warning with following snippet:

```javascript
(function() {
    'use strict';
    var name = 'hello';

}());
```

```
  1:1  warning  Use the global form of 'use strict'  strict
  2:5  warning  Use the global form of 'use strict'  strict

✖ 2 problems (0 errors, 2 warnings)
```

A very simple `iife`, which should be treated as `Everywhere else, "safe" mode corresponds to "function" mode and enforces strict mode directives inside top-level functions`.

But warning was shown, am i wrong again?

**Comment by btmills** [comment]
Are you still using the same config from above?

**Comment by nzakas** [comment]
As far as I can tell, there's no bug here. In the first instance, you were resetting `globalReturn` to false, which made the rule enforce function strict mode. In the second, the safe setting is now global and you have it in a function.

**Comment by leftstick** [comment]
Hi @btmills, i am still using the above config except `globalReturn` was enabled, as following:

```javascript
{
    "parserOptions":{
        "ecmaVersion": 6,
        "sourceType": "script",
        "ecmaFeatures": {
            "globalReturn": true,
            "impliedStrict": false,
            "jsx": false,
            "experimentalObjectRestSpread": true
        }
    },
    "env": {
        "shared-node-browser": true,
        "commonjs": true,
        "node": true,
        "es6": true,
        "mocha": true,
        "jquery": true
    },
    "rules": {
        "strict": [1, "safe"]
    }
}
```

Let me describe my issue a bit. I have two flavors coding style in my project:

1. written in `commonjs` module
2. legacy code, written in `iife`

When i saw the [strict-safe-doc](http://eslint.org/docs/rules/strict#safe-default), i thought this mode might provide me some "smart" way to determine what style the code is, and display warning properly like i expected:

*  the `use strict` directive should be placed in `global` location once the code is written in `commonjs` module
* the `use strict` directive should be placed in `function` location once the code is written in `iife`

But according to @nzakas 's explanation, there is no such a thing. The key to display warning is how i set `globalReturn`. So why we have `safe` mode, what is the original thought/purpose?

Thanks very much

**Comment by nzakas** [comment]
Please see https://github.com/eslint/eslint/issues/3306 for background
